fenix/app/src/main/java/org/mozilla/fenix/messaging/state/MessagingMiddleware.kt

```diff
@@ -37,7 +38,9 @@ class MessagingMiddleware(
             is Restore -> {
                 coroutineScope.launch {
                     val messages = controller.getMessages()
-                    context.store.dispatch(UpdateMessages(messages))
+                    MainScope().launch {
```

Comment:
🚗 drive-by:

🤔 I'm not sure what the fix is, but I don't think we should be doing this - it looks like a code smell to me. My interpretation of:

> Note that this method should only ever be called from a [Middleware] and the calling thread.
doesn't mean it has to be the main thread

Reply:
Happy to Rubber Duck or pair if it helps 🙂

Reply:
@rocketsroger could you clarify why w need the `MainScope`? 

Checking at stack trace,[ it looks like crash is coming form ](https://crash-stats.mozilla.org/report/index/d881258e-ca72-4f99-a2e7-7773a0240226#tab-details) `org.mozilla.fenix.messaging.state.MessagingMiddleware$onMessagedDisplayed$1`

Reply:
Happy to also pair on the solution :)

Reply:
The crash is coming from our own [assertion](https://github.com/mozilla-mobile/firefox-android/blob/86bd17c82b227f3a7e538e7a469d82a57541082e/android-components/components/lib/state/src/main/java/mozilla/components/lib/state/internal/StoreThreadFactory.kt#L53-L56) 
``` 
`Expected `store` thread, but running on thread `DefaultDispatcher-worker-7`. Leaked MiddlewareContext or did you mean to use `MiddlewareContext.store.dispatch`?
```

 Indicating that we should be using the store thread.

Reply:
> Indicating that we should be using the store thread.

That's what I saw too.  The Store's thread is Main so that's why I try to dispatch everything on main.  But it does feel like a hack.  I don't know if we have a better solution unless we revert Jame's changes.

Reply:
> doesn't mean it has to be called from the main thread?

The store was created on main thread so the dispatch have to be called on the main thread.  Maybe I can let the caller pass in that.

Reply:
Alternative we could also do:

```kotlin

                coroutineScope.launch {
                    val messages = controller.getMessages()
                    withContext(Dispatchers.Main) {
                        context.store.dispatch(UpdateMessages(messages))
                    }
                } 
```